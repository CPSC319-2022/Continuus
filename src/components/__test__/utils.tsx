import React, { type PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import { configureStore, type PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";

import type { AppStore, RootState } from "~/redux/store";
import { reducer } from "~/redux/slices";
// As a basic setup, import your same slice reducers

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({ reducer: reducer, preloadedState }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  // eslint-disable-next-line @typescript-eslint/ban-types
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
