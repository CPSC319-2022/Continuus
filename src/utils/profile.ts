/**
 * Returns a path to the profile page for the given profile. If the argument is undefined, redirect to the home page.
 *
 * @param profile the user id of the account
 *
 * @return the profile path if the profile is a string, otherwise the home path
 */
export const userPathToProfile = (profile: string | undefined): string => {
    if (profile === undefined) {
        return '/';
    }

    return '/profile/' + encodeURIComponent(profile);
}
