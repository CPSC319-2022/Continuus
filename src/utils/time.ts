export const timeAgo = (inputDate: Date): string => {
  const now = new Date();
  const diffInSeconds = Math.max(Math.floor(
    (now.getTime() - inputDate.getTime()) / 1000
  ), 0);

  if (diffInSeconds < 60) {
    return `${diffInSeconds} ${diffInSeconds === 1 ? "second" : "seconds"} ago`;
  }

  const diffInMinutes = Math.floor(diffInSeconds / 60);
  if (diffInMinutes < 60) {
    return `${diffInMinutes} ${diffInMinutes === 1 ? "minute" : "minutes"} ago`;
  }

  const diffInHours = Math.floor(diffInMinutes / 60);
  if (diffInHours < 24) {
    return `${diffInHours} ${diffInHours === 1 ? "hour" : "hours"} ago`;
  }

  const diffInDays = Math.floor(diffInHours / 24);
  if (diffInDays < 30) {
    return `${diffInDays} ${diffInDays === 1 ? "day" : "days"} ago`;
  }

  const diffInMonths = Math.floor(diffInDays / 30);
  if (diffInMonths < 12) {
    return `${diffInMonths} ${diffInMonths === 1 ? 'month' : 'months'} ago`;
  }

  const diffInYears = Math.floor(diffInMonths / 12);
  return `${diffInYears} ${diffInYears === 1 ? 'year' : 'years'} ago`;
};
