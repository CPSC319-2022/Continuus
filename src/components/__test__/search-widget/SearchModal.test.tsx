import { describe, it, expect, beforeEach, vi } from "vitest";
import { cleanup, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { renderWithProviders } from "../utils";

const props = {
  isOpen: true,
  setIsOpen: vi.fn(),
};

const userResults = [
  {
    id: "id1",
    name: "test1",
    image: "test1",
  },
  {
    id: "id2",
    name: "test2",
    image: "test2",
  },
  {
    id: "id3",
    name: "test3",
    image: "test3",
  },
];

const date = new Date();

const blogResults = [
  {
    id: "id1",
    title: "title1",
    updatedAt: date,
    createdAt: date,
    content: "test",
    user: {
      name: "test1",
      image: "image",
    },
  },
  {
    id: "id2",
    title: "title2",
    updatedAt: date,
    createdAt: date,
    content: "test",
    user: {
      name: "test2",
      image: "image",
    },
  },
];

const setup = (userResults: any = [], blogResults: any = []) => {
  vi.doMock("~/utils/api", () => ({
    api: {
      user: {
        searchUsers: {
          useQuery: () => ({
            data: userResults,
            isLoading: false,
          }),
        },
      },
      blogPost: {
        search: {
          useQuery: () => ({
            data: blogResults,
            isLoading: false,
          }),
        },
      },
    },
  }));
};

describe("Search: Snapshot", () => {
  beforeEach(() => {
    cleanup();
    vi.clearAllMocks();
    vi.resetModules();
  });

  beforeEach(() => {
    vi.useRealTimers();
  });

  it("render with no results", async () => {
    setup();
    const { SearchModal } = await import(
      "~/components/search-widget/SearchModal"
    );
    const body = renderWithProviders(<SearchModal {...props} />).baseElement;
    expect(body).toMatchSnapshot();
  });

  it("render with blog results", async () => {
    setup([], blogResults);
    const { SearchModal } = await import(
      "~/components/search-widget/SearchModal"
    );
    const body = renderWithProviders(<SearchModal {...props} />);
    const input = screen.getByTestId("search-input");
    await userEvent.type(input, "hello");
    await waitFor(() => {
      expect(screen.getAllByTestId("search-posts")).toHaveLength(2);
    });
    expect(body.baseElement).toMatchSnapshot();
  });

  it("render with user results", async () => {
    setup(userResults, []);
    const { SearchModal } = await import(
      "~/components/search-widget/SearchModal"
    );
    const body = renderWithProviders(<SearchModal {...props} />);
    const input = screen.getByTestId("search-input");
    await userEvent.type(input, "@test");
    await waitFor(() => {
      expect(screen.getAllByTestId("search-users")).toHaveLength(3);
    });
    expect(body.baseElement).toMatchSnapshot();
  });
});
