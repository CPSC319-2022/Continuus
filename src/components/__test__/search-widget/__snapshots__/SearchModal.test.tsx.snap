// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Search: Snapshot > render with blog results 1`] = `
<body
  class="ReactModal__Body--open"
>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div />
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div
    class="ReactModalPortal"
  >
    <div
      class="ReactModal__Overlay ReactModal__Overlay--after-open fixed inset-0 z-20 bg-black/75"
    >
      <div
        aria-modal="true"
        class="ReactModal__Content ReactModal__Content--after-open absolute top-1/2 left-1/2 z-40 h-[40rem] w-11/12 -translate-x-1/2 -translate-y-1/2 overflow-y-auto rounded-md border-slate-500 bg-white outline-none md:w-1/2"
        role="dialog"
        tabindex="-1"
      >
        <div
          class="p-8"
          data-testid="search-modal"
        >
          <input
            class="input-bordered sticky mb-1 w-full rounded-sm border-[1px] p-2"
            data-testid="search-input"
            placeholder="Search for a blog post or user..."
            type="text"
            value="hello"
          />
          <div
            class="my-2 flex rounded-md border-b p-3 transition-all hover:cursor-pointer hover:bg-gray-200"
            data-testid="search-posts"
          >
             
            <img
              alt="avatar"
              class="inline-block aspect-square rounded-[50%] self-center"
              src="https://i.stack.imgur.com/34AD2.jpg"
              style="width: 2rem;"
            />
            <div
              class="ml-3"
            >
              <p
                class="font-bold"
              >
                title1
              </p>
              <p
                class="text-xs"
              >
                test
              </p>
            </div>
          </div>
          <div
            class="my-2 flex rounded-md border-b p-3 transition-all hover:cursor-pointer hover:bg-gray-200"
            data-testid="search-posts"
          >
             
            <img
              alt="avatar"
              class="inline-block aspect-square rounded-[50%] self-center"
              src="https://i.stack.imgur.com/34AD2.jpg"
              style="width: 2rem;"
            />
            <div
              class="ml-3"
            >
              <p
                class="font-bold"
              >
                title2
              </p>
              <p
                class="text-xs"
              >
                test
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Search: Snapshot > render with no results 1`] = `
<body
  class="ReactModal__Body--open"
>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div />
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div
    class="ReactModalPortal"
  >
    <div
      class="ReactModal__Overlay fixed inset-0 z-20 bg-black/75"
    >
      <div
        aria-modal="true"
        class="ReactModal__Content absolute top-1/2 left-1/2 z-40 h-[40rem] w-11/12 -translate-x-1/2 -translate-y-1/2 overflow-y-auto rounded-md border-slate-500 bg-white outline-none md:w-1/2"
        role="dialog"
        tabindex="-1"
      >
        <div
          class="p-8"
          data-testid="search-modal"
        >
          <input
            class="input-bordered sticky mb-1 w-full rounded-sm border-[1px] p-2"
            data-testid="search-input"
            placeholder="Search for a blog post or user..."
            type="text"
            value=""
          />
          <div
            class="flex h-[35rem] flex-col justify-center align-middle"
          >
            <p
              class="self-center"
            >
              Start typing to search for a blog post!
               
            </p>
            <p
              class="self-center"
            >
              Type @ at the beginning to search for a user.
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Search: Snapshot > render with user results 1`] = `
<body
  class="ReactModal__Body--open"
>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div />
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div
    class="ReactModalPortal"
  >
    <div
      class="ReactModal__Overlay ReactModal__Overlay--after-open fixed inset-0 z-20 bg-black/75"
    >
      <div
        aria-modal="true"
        class="ReactModal__Content ReactModal__Content--after-open absolute top-1/2 left-1/2 z-40 h-[40rem] w-11/12 -translate-x-1/2 -translate-y-1/2 overflow-y-auto rounded-md border-slate-500 bg-white outline-none md:w-1/2"
        role="dialog"
        tabindex="-1"
      >
        <div
          class="p-8"
          data-testid="search-modal"
        >
          <input
            class="input-bordered sticky mb-1 w-full rounded-sm border-[1px] p-2"
            data-testid="search-input"
            placeholder="Search for a blog post or user..."
            type="text"
            value="@test"
          />
          <a
            class="my-2 flex rounded-md border-b p-3 transition-all hover:cursor-pointer hover:bg-gray-200"
            data-testid="search-users"
            href="/profile/id1"
          >
             
            <img
              alt="avatar"
              class="inline-block aspect-square rounded-[50%] self-center"
              src="https://i.stack.imgur.com/34AD2.jpg"
              style="width: 2rem;"
            />
            <div
              class="ml-3 flex"
            >
              <p
                class="self-center"
              >
                <mark
                  class="bg-sky-200 dark:bg-amber-200"
                >
                  test
                </mark>
                1
              </p>
            </div>
          </a>
          <a
            class="my-2 flex rounded-md border-b p-3 transition-all hover:cursor-pointer hover:bg-gray-200"
            data-testid="search-users"
            href="/profile/id2"
          >
             
            <img
              alt="avatar"
              class="inline-block aspect-square rounded-[50%] self-center"
              src="https://i.stack.imgur.com/34AD2.jpg"
              style="width: 2rem;"
            />
            <div
              class="ml-3 flex"
            >
              <p
                class="self-center"
              >
                <mark
                  class="bg-sky-200 dark:bg-amber-200"
                >
                  test
                </mark>
                2
              </p>
            </div>
          </a>
          <a
            class="my-2 flex rounded-md border-b p-3 transition-all hover:cursor-pointer hover:bg-gray-200"
            data-testid="search-users"
            href="/profile/id3"
          >
             
            <img
              alt="avatar"
              class="inline-block aspect-square rounded-[50%] self-center"
              src="https://i.stack.imgur.com/34AD2.jpg"
              style="width: 2rem;"
            />
            <div
              class="ml-3 flex"
            >
              <p
                class="self-center"
              >
                <mark
                  class="bg-sky-200 dark:bg-amber-200"
                >
                  test
                </mark>
                3
              </p>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</body>
`;
